synapse/handlers/federation_event.py

```diff
@@ -1939,3 +1949,33 @@ def _sanity_check_event(self, ev: EventBase) -> None:
                 len(ev.auth_event_ids()),
             )
             raise SynapseError(HTTPStatus.BAD_REQUEST, "Too many auth_events")
+
+    def strip_unsigned_values(self, pdu: EventBase, event_dict: Dict) -> EventBase:
+        """
+        Strip any unsigned values unless specifically allowed, as defined by the whitelist.
+
+        pdu: the event to strip values from
+        event_dict: a dict of values derived from the pdu
+        """
+        unsigned = event_dict.get("unsigned")
+
+        if pdu.type == EventTypes.Member:
+            whitelist = ["knock_room_state", "invite_room_state", "age"]
+        else:
+            whitelist = ["age"]
+
+        filtered_unsigned = {}
+
+        # Unsigned should never be None by the time we are here but mypy doesn't know that
+        assert unsigned is not None
+        for k, v in unsigned.items():
+            if k in whitelist:
+                filtered_unsigned[k] = v
+
+        event_dict["unsigned"] = filtered_unsigned
```

Comment:
This mutates the `event_dict` in place. Assuming that's intentional, I think we should note this in the docstring.

(Note that EventBase.get_dict() returns a copy of the event's internal dict, so... we're not mutating the original event. I guess I was just a bit surprised that operate in two modes simulatneously: we mutate the `event_dict` parameter but build a new EventBase.)

Reply:
Right I can see how that would be a little ...unclean... What might be a more graceful way to handle this? I think it makes sense to build a new EventBase rather than mutating the original event, would it make sense to make a copy of the `event_dict` parameter (even though it is already a copy, as you mentioned) and use that to build the EventBase rather than the mutated parameter?